### DISCUSSION TITLE
Permission problems when checking out code as part of GitHub action

### DISCUSSION THREAD

**Discussion by jeltsch** [community/community#25606] [Actions]
<p>I’m trying to use an image from Docker Hub, <a href="https://hub.docker.com/r/makarius/isabelle" title="Official Isabelle Docker distribution" rel="noopener nofollow ugc"><code>makarius/isabelle</code></a> , for automated builds. However, if usage of this image is enabled, checking out my repository with <code>actions/checkout@v2</code> fails with the error <code>EACCES: permission denied, open '/__w/⟨repository-name⟩/⟨repository-name⟩/⟨uuid⟩.tar.gz'</code>. The repository is downloaded using the GitHub REST API.</p>
<p>A minimal workflow that triggers this error is as follows:</p>
<pre><code class="lang-auto">name: Automated checkout

on: [push]

jobs:
  checkout:
    runs-on: ubuntu-latest
    container: makarius/isabelle
    steps:
      - name: Check out repository
        uses: actions/checkout@v2
</code></pre>
<p>Unfortunately, the above error message, while mentioning a path, does not tell, on which machine the respective directories should exist and what permissions they should have.</p>
<p>From its documentation, I cannot tell, what <code>actions/checkout@v2</code> is trying to do precisely. Should the virtual host write the data into the container but doesn’t have write permissions? Or should the container try to copy the data from the virtual host onto its own disk but can’t either read or write the data?</p>
<p>Are there any constraints a docker image to be used with <code>actions/checkout@v2</code> has to fulfill? If yes, where are these documented?</p>
<p>The particular image I’m trying to employ uses a non-privileged user for its entry point. Could this be the problem?</p>

**Comment by jeltsch** [comment]
<p>I’ve extended the workflow, adding <code>run: ls -laR /__w</code> before the checkout. Now I see that the directory <code>/__w</code> indeed exists inside the container; so apparently the host system manages to create or mount it there.</p>
<p>The problem is that this directory and everything in it is created for user <code>1001</code> and group <code>121</code>, which don’t exist inside the container. Apparently, <code>actions/checkout@v2</code> just assumes that the container uses the same user database as the host, which is quite a bold assumption, in my opinion. I guess the containers provided by GitHub Actions have this property, but since the <code>checkout</code> action has special support for using images from Docker Hub, I’m a bit surprised that such a constraint is imposed (and apparently not documented).</p>
<p>What is a simple way to fix that?</p>
<p>My current idea is that the only safe approach would be to use a Docker image derived from one of GitHub’s own images, like <code>ubuntu-latest</code>, and have this image contain the Isabelle software. However, can I create and store such an image within GitHub Actions? Putting it on Docker Hub doesn’t seem like a good idea to me, since it would depend on one of GitHub’s images, which are moving targets, from what I understand.</p>

**Comment by jeltsch** [comment, marked as answer]
<p>I’m meanwhile convinced that the container’s entry point must have <code>root</code> privileges. This constraint is documented in the <a href="https://docs.github.com/en/actions" title="GitHub Actions Documentation" rel="noopener nofollow ugc">GitHub Actions Documentation</a> in the section “Creating actions”. My approach doesn’t involve creating a custom action, which is why I didn’t look at that section initially. Apparently, the requirement of having <code>root</code> privileges also applies to the situation where you don’t implement your own action, but unfortunately the documentation doesn’t seem to mention that.</p>

**Comment by 64J0** [comment]
Facing the same problem. I would like to avoid using the root user in this environment.
